### PR TITLE
added ability to normalize large datasets in memory

### PR DESCRIPTION
**Pull request by evanfeinberg**
This PR adds a "preprocessing" section which is inspired by the approach of geometric clustering in `MSMBuilder`: take standard `sklearn` methods and expand them to handle `list`s of `np.array`s. This is more convenient, as nearly all `sklearn` methods assume that `X` is a single `np.array` whereas MD data is mostly stored in lists of arrays. 

Large featurized [MD] datasets are often large compared to total RAM available on a system. If size of the MD dataset is comparable to the total RAM available, then computing the mean and variance with standard `numpy` functions can be tricky. Since concatenating the data essentially doubles the RAM that needs to be used, the usual way of computing mean and variance can in turn make the user exceed the memory limits of the system.

To begin, I implemented a numerically stable method described in https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Compensated_variant. In addition to the normal `fit`, `transform`, and `fit_transform` methods, I added `fit_transform_in_place` and `transform_in_place`, so one can pass in un-normalized features and compute normalized features without adding to the memory burden of one's computation. 

Some to-do's: add whitening and other preprocessing capabilities from `sklearn` to `msmbuilder.preprocessing`. 

Comments or criticisms are welcome. Thank you to @cxhernandez  in particular for his help with this PR.

### CODE REVIEW

**Comment by mpharrigan** [comment]
What is `msmbuilder/preprocessing/._normalize.py`? Can you remove it and rebase so it doesn't take up space in the history

File: msmbuilder/preprocessing/base.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
+# Author: Matthew Harrigan <matthew.p.harrigan@gmail.com>
```
Comment (mpharrigan) [inline msmbuilder/preprocessing/base.py]: lol I don't *think* I wrote this :)

File: msmbuilder/preprocessing/normalize.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
+# Author: Evan N. Feinberg <efein19@gmail.com>
+# Contributors: Christian Schwantes <schwancr@gmail.com>m Robert McGibbon <rmcgibbo@gmail.com>, Kyle A. Beauchamp  <kyleabeauchamp@gmail.com>
```
Comment (mpharrigan) [inline msmbuilder/preprocessing/normalize.py]: Hard-wrap this line

**Comment by mpharrigan** [comment]
I don't really understand this PR. The end goal is to remove mean and normalize variance for feature-trajectories (aka 2d numpy arrays)? Can you really not fit a featurized trajectory in ram?

**Comment by mpharrigan** [comment]
Back of the envelope, one milisecond saved every nanosecond with 1000 features only takes 4gib of ram. This is the same example I used in #768 to argue for just loading everything into ram

This is also pretty conservative. For normalizing, you can load in one trajectory at a time (using `dataset`) which will be << 1 ms or save less frequently than every 1 ns

**Comment by mpharrigan** [comment]
Oh I guess you have to consider *all* the data when doing mean+std calculations

File: msmbuilder/preprocessing/normalize.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        `partial_fit()` updates the fit with new data, and is suitable for
+         online learning.
+
+    Note: Numerically stable method of computing variance adapted from:
+    https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance
+    """
+
+    def __init__(self):
+
+        """
```
Comment (mpharrigan) [inline msmbuilder/preprocessing/normalize.py]: `__ini__` generally doesn't have a docstring. Put this in the class docstring or as line comments `# (these)`

File: msmbuilder/preprocessing/normalize.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+        self.k = len(sequences)
+        self.n = 0.
+        self.sum1 = 0.
+        self.sum2 = 0.
+        self.sum3 = 0.
+
+        #First Pass
+        for sequence in sequences:
+            self.n += float(np.shape(sequence)[0])
```
Comment (mpharrigan) [inline msmbuilder/preprocessing/normalize.py]: Why is `n` a float?

File: msmbuilder/preprocessing/normalize.py (hunk 2 above)
Comment (mpharrigan) [inline msmbuilder/preprocessing/normalize.py]: I also suspect not all these people contributed

**Comment by mpharrigan** [comment]
Why do you include ` msmbuilder/preprocessing/base.py`?

**Comment by mpharrigan** [comment]
Don't you need to implement `partial_transform()` in the Normalize class?

**Comment by mpharrigan** [comment]
This also needs a suite of tests :)

File: msmbuilder/preprocessing/base.py (hunk 1 above)
Comment (evanfeinberg) [inline msmbuilder/preprocessing/base.py]: You indeed wrote the vast majority of it!

**Comment by evanfeinberg** [comment]
Yes this is precisely why this PR is necessary. Suppose your data, `sequence` takes up 10 GB in memory and your system only has 16 GB of RAM (both very sensible figures). Computing mean and (especially) variance requires seeing all the data. Calling `np.concatenate` on `sequence` is essentially a copy. If you want to keep around `sequence` and `sequence_concatenated`, this will push memory usage beyond the 16 GB limit. In addition, for transforming, making a `sequence_normalized` separate from `sequence` would similarly ~double memory requirements. Therefore a `transform_in_place` is useful.

**Comment by evanfeinberg** [comment]
Doesn't `msmbuilder/preprocessing/base.py` obviate implementing `partial_transform` ?

File: msmbuilder/preprocessing/normalize.py (hunk 3 above)
Comment (evanfeinberg) [inline msmbuilder/preprocessing/normalize.py]: Agreed, will be removed.

File: msmbuilder/preprocessing/normalize.py (hunk 4 above)
Comment (evanfeinberg) [inline msmbuilder/preprocessing/normalize.py]: So when we divide sum of squares by `n` Python doesn't get confused and cast the sum to an `int` which would be bad. Unless it automatically casts `n` to a float underneath the hood in which case we're ok?

**Comment by evanfeinberg** [comment]
Also how can I add tests?

File: msmbuilder/preprocessing/normalize.py (hunk 2 above)
Comment (evanfeinberg) [inline msmbuilder/preprocessing/normalize.py]: Oops you're totally right, corrected.

File: msmbuilder/preprocessing/normalize.py (hunk 4 above)
Comment (mpharrigan) [inline msmbuilder/preprocessing/normalize.py]: python2 will truncate it to int (dumb python 2!). python3 will be nice. You can get the python3 behavior by putting

```python
from __future__ import division
```

at the top

**Comment by mpharrigan** [comment]
> Also how can I add tests?

Check out the files in `msmbuilder/tests/`.

I would recommend adding a new file called `test_normalize.py`. In it, you can put a couple functions whose name start with `test_` that each do one little thing and make an assertion at the end. These functions will all be run on travis, and if anything raises an exception or assertion error, that particular test will fail. You can use the script `runtests.py` in the project root to re-build and run the tests. You can also run `nosetests msmbuilder` to test your currently installed copy. To just do a particular test file, you can do `nosetests msmbuilder.tests.test_normalize`. You can't run this program from the source directory! or it will try to import the local files instead of the built+installed copy

**Comment by mpharrigan** [comment]
> Doesn't msmbuilder/preprocessing/base.py obviate implementing partial_transform ?

You don't inherit from any of the classes there. As it stands, I don't see you using any of the functionality form base.py. If that's just a copy-paste of the file in decomposition... don't. import from decomposition. But I'm pretty sure you're not using and shouldn't use any of that functionality. The whole point of this PR is to avoid concatenating things

**Comment by mpharrigan** [comment]
> Suppose your data, sequence takes up 10 GB in memory and your system only has 16 GB of RAM

Can you address my envelop calculation

> one milisecond saved every nanosecond with 1000 features only takes 4gib of ram

**Comment by evanfeinberg** [comment]
Wow, WIP #768 looks really cool! Unfortunately, if you have several thousand features, as I often do, 1000 features / 4GB --> 5000 features / 20 GB easily. For instance, suppose one has 1 ms of trajectory subsampled at every 1 ns. Measuring ~2500 `closest-heavy` and `CA` distances translates to 5000 features and 20 GB of data. On a lot of hardware this is a nightmare.

**Comment by evanfeinberg** [comment]
With regards to base.py both `transform()` and `fit_transform()` are only implemented in base.py, not in normalize.py. I figured having base.py in the class would be useful to other `preprocessing` classes we will end up having. But yes right now it is mostly a pared-down copy paste.

File: msmbuilder/preprocessing/normalize.py (hunk 4 above)
Comment (evanfeinberg) [inline msmbuilder/preprocessing/normalize.py]: cool, done!

**Comment by mpharrigan** [comment]
> With regards to base.py both transform() and fit_transform() are only implemented in base.py, not in normalize.py.

You don't inherit from it anywhere. The code in base.py isn't used

**Comment by rmcgibbo** [comment]
I just saw this on my phone and didn't read rhe whole thing... but IIRC in
mdtraj I implemented running estimators of the mean and variance that take
a single pass through the data and are in C or cython so they should be
both fast and stable. Could be useful for this PR.

**Comment by rmcgibbo** [comment]
Here's the code in mdtraj for an online (running) calculation of the first couple moments of a bunch of data. It doesn't look like I've wrapped this from cython, but if it would be useful I could do that.

https://github.com/mdtraj/mdtraj/blob/master/mdtraj/geometry/src/moments.cpp

**Comment by evanfeinberg** [comment]
Oh cool, it looks like I ended up duplicating the same algorithm implementation from that great Wiki page on variance computation. @rmcgibbo is this function callable from mdtraj Python API easily?

**Comment by cxhernandez** [comment]
> It doesn't look like I've wrapped this from cython, but if it would be useful I could do that.

^ @evanfeinberg

**Comment by msultan** [comment]
can we get this into master? Seems like a very useful addition

**Comment by mpharrigan** [comment]
There are still a couple of issues with the code that need to be addressed

1. docstrings
2. completely extraneous file that I think @evanfeinberg thinks he is using but actually is not
3. tests

And introducing a whole new API not found in sklearn is something I would like to give some thought to.

1. is this the best name? can we do this as an init arg instead?
2. how does sklearn handle these issues? can we do something like that?

It also looks like there was a conversation about using the c implementation to make this speedy, so @evanfeinberg and @rmcgibbo need to decide whether they're going to do that

**Comment by mpharrigan** [comment]
For the second point 2, perhaps interested parties can look into this

http://scikit-learn.org/stable/modules/preprocessing.html

http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html#sklearn.preprocessing.StandardScaler

I'd be more comfortable using a thin wrapper around sklearn. At the very least, we should copy their api design. It's essentially the second part of my second number 1: they use an init arg `copy`. When set to false, it does stuff in place

**Comment by cxhernandez** [comment]
I've implemented point 2 in https://github.com/msmbuilder/msmbuilder/pull/796.

@evanfeinberg: I still think it's worth checking if @rmcgibbo's code runs faster than the sklearn version.

**Comment by mpharrigan** [comment]
I'm going to close this because of #796 . Feel free to file additional PRs for performance enhancements if necessary
